### ISSUE TITLE
Regression: Idris custom Setup.hs breaks

### ISSUE THREAD

**Issue by mietek** [haskell/cabal#2342]
Something changed between 1.20.0.3 and 1.22.0.0, causing the Idris custom [`Setup.hs`](https://github.com/idris-lang/Idris-dev/blob/master/Setup.hs) to include invalid values.

Looking at `dist/build/autogen/Paths_idris.hs`, both `libdir` and `datadir` seem wrong:
```
libdir     = "/app/lib/$abi/idriskey"
datadir    = "/app/share/$abi/idris-0.9.15.1"
```

See also https://github.com/idris-lang/Idris-dev/issues/1836.

**Comment by luite** [comment]
Are you using `cabal-install` 1.22 with Cabal library 1.20? (for example in a sandbox where user db Cabal library is not visible)

**Comment by mietek** [comment]
Indeed, a sandbox.

**Comment by luite** [comment]
And does the problem go away if you install Cabal 1.22 in the sandbox?

**Comment by luite** [comment]
A bit more background info for if you want to investigate: `$abi` is a new variable in 1.22 that should uniquely identify the compiler in terms of binary compatibility.

What I think happened is that you initialized the sandbox with 1.22, then configured with `cabal-install`, which had to fall back to 1.20 because of the external setup method forced by the custom setup type and no visible 1.22 Cabal library. The `$abi` variable is used in the generated `cabal.sandbox.config`, but 1.20 does not recognize it.

I wonder how this can be dealt with in a more robust fashion. In the global `~/.cabal/config` all the subdirs are commented out by default, so this problem only shows up if users override the default values themselves. In the generated `cabal.sandbox.config` they're not commented out, which makes upgrading to new defaults trickier.

I wonder if it is really necessary for the values, other than `prefix` to be uncommented in `cabal.sandbox.config`? @23Skidoo , @dcoutts thoughts?

**Comment by mietek** [comment]
@luite Please note _cabal-install_ 1.22.0.0 reports it’s using version 1.22.0.0 of the Cabal library even when there is no Cabal library 1.22.0.0 registered with `ghc-pkg`:
```
$ cabal --version
cabal-install version 1.22.0.0
using version 1.22.0.0 of the Cabal library
 
$ ghc-pkg list
/app/ghc/lib/ghc-7.8.4/package.conf.d
   Cabal-1.18.1.5
   array-0.5.0.0
   base-4.7.0.2
   bin-package-db-0.0.0.0
   binary-0.7.1.0
   bytestring-0.10.4.0
   containers-0.5.5.1
   deepseq-1.3.0.2
   directory-1.2.1.0
   filepath-1.3.0.2
   ghc-7.8.4
   ghc-prim-0.3.1.0
   haskeline-0.7.1.2
   haskell2010-1.1.2.0
   haskell98-2.0.0.3
   hoopl-3.10.0.1
   hpc-0.6.0.1
   integer-gmp-0.5.1.0
   old-locale-1.0.0.6
   old-time-1.1.0.2
   pretty-1.1.1.1
   process-1.2.0.0
   rts-1.0
   template-haskell-2.9.0.0
   terminfo-0.4.0.0
   time-1.4.2
   transformers-0.3.0.0
   unix-2.7.0.1
   xhtml-3000.2.1
```

**Comment by luite** [comment]
Yes `cabal-install` just reports the `Cabal` version it was built with, not the one available for building custom `Setup.hs`

**Comment by ttuegel** [comment]
Especially when we are in a sandbox, we should seriously consider having cabal-install-x.y.* require Cabal-x.y.* for Custom-type builds.

**Comment by BardurArantsson** [comment]
I can see that this would be a problem. Can we please recast this issue (or create a new one) which specifies the *actual* problem up front (and esp. in the summary/title)?

**Comment by mietek** [comment]
@ttuegel, is this still relevant?
